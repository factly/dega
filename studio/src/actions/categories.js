import axios from 'axios';
import {
  ADD_CATEGORIES,
  ADD_CATEGORIES_REQUEST,
  SET_CATEGORIES_LOADING,
  RESET_CATEGORIES,
  CATEGORIES_API,
  GET_CATEGORY,
  UPDATE_CATEGORY,
} from '../constants/categories';
import { addErrorNotification, addSuccessNotification } from './notifications';
import { addMedia } from './media';
import getError from '../utils/getError';

// action to fetch all categories
export const getCategories = (query) => {
  return (dispatch, getState) => {
    const currentSpaceID = getState().spaces?.selected;
    if (currentSpaceID === 0) {
      return;
    }
    dispatch(loadingCategories());
    return axios
      .get(CATEGORIES_API, {
        params: query,
      })
      .then((response) => {
        dispatch(
          addMedia(
            response.data.nodes
              .filter((category) => category.medium)
              .map((category) => category.medium),
          ),
        );
        dispatch(
          addCategoriesList(
            response.data.nodes.map((category) => {
              return { ...category, medium: category.medium?.id };
            }),
          ),
        );
        dispatch(
          addCategoriesRequest({
            data: response.data.nodes.map((item) => item.id),
            query: query,
            total: response.data.total,
          }),
        );
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopCategoriesLoading()));
  };
};

// action to fetch category by id
export const getCategory = (id) => {
  return (dispatch) => {
    dispatch(loadingCategories());
    return axios
      .get(CATEGORIES_API + '/' + id)
      .then((response) => {
        if (response.data.medium) dispatch(addMedia([response.data.medium]));

        dispatch(addCategory(GET_CATEGORY, { ...response.data, medium: response.data.medium?.id }));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopCategoriesLoading()));
  };
};

// action to create category
export const createCategory = (data) => {
  return (dispatch) => {
    dispatch(loadingCategories());
    return axios
      .post(CATEGORIES_API, data)
      .then(() => {
        dispatch(resetCategories());
        dispatch(addSuccessNotification('Category created'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      });
  };
};

// action to update category by id
export const updateCategory = (data) => {
  return (dispatch) => {
    dispatch(loadingCategories());
    return axios
      .put(CATEGORIES_API + '/' + data.id, data)
      .then((response) => {
        if (response.data.medium) dispatch(addMedia([response.data.medium]));

        dispatch(
          addCategory(UPDATE_CATEGORY, { ...response.data, medium: response.data.medium?.id }),
        );
        dispatch(addSuccessNotification('Category updated'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopCategoriesLoading()));
  };
};

// action to delete category by id
export const deleteCategory = (id) => {
  return (dispatch) => {
    dispatch(loadingCategories());
    return axios
      .delete(CATEGORIES_API + '/' + id)
      .then(() => {
        dispatch(resetCategories());
        dispatch(addSuccessNotification('Category deleted'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      });
  };
};

export const addCategories = (categories) => {
  return (dispatch) => {
    dispatch(
      addMedia(categories.filter((category) => category.medium).map((category) => category.medium)),
    );
    dispatch(
      addCategoriesList(
        categories.map((category) => {
          return { ...category, medium: category.medium?.id };
        }),
      ),
    );
  };
};

export const loadingCategories = () => ({
  type: SET_CATEGORIES_LOADING,
  payload: true,
});

export const stopCategoriesLoading = () => ({
  type: SET_CATEGORIES_LOADING,
  payload: false,
});

export const addCategory = (type, payload) => ({
  type,
  payload,
});

export const addCategoriesList = (data) => ({
  type: ADD_CATEGORIES,
  payload: data,
});

export const addCategoriesRequest = (data) => ({
  type: ADD_CATEGORIES_REQUEST,
  payload: data,
});

export const resetCategories = () => ({
  type: RESET_CATEGORIES,
});
