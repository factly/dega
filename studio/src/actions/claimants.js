import axios from 'axios';
import {
  ADD_CLAIMANTS,
  ADD_CLAIMANTS_REQUEST,
  SET_CLAIMANTS_LOADING,
  RESET_CLAIMANTS,
  CLAIMANTS_API,
  UPDATE_CLAIMANT,
  GET_CLAIMANT,
} from '../constants/claimants';
import { addErrorNotification, addSuccessNotification } from './notifications';
import { addMedia } from './media';
import getError from '../utils/getError';

// action to fetch all claimants
export const getClaimants = (query) => {
  return (dispatch) => {
    dispatch(loadingClaimants());
    return axios
      .get(CLAIMANTS_API, {
        params: query,
      })
      .then((response) => {
        dispatch(
          addMedia(
            response.data.nodes
              .filter((claimant) => claimant.medium)
              .map((claimant) => claimant.medium),
          ),
        );
        dispatch(
          addClaimantsList(
            response.data.nodes.map((claimant) => {
              claimant.description = {
                json: claimant.description,
                html: claimant.html_description,
              };
              return { ...claimant, medium: claimant.medium?.id };
            }),
          ),
        );
        dispatch(
          addClaimantsRequest({
            data: response.data.nodes.map((item) => item.id),
            query: query,
            total: response.data.total,
          }),
        );
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopClaimantsLoading()));
  };
};

// action to fetch claimant by id
export const getClaimant = (id) => {
  return (dispatch) => {
    dispatch(loadingClaimants());
    return axios
      .get(CLAIMANTS_API + '/' + id)
      .then((response) => {
        if (response.data.medium) dispatch(addMedia([response.data.medium]));
        response.data.description = {
          json: response.data.description,
          html: response.data.html_description,
        };
        dispatch(addClaimant(GET_CLAIMANT, { ...response.data, medium: response.data.medium?.id }));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopClaimantsLoading()));
  };
};

// action to create claimant
export const createClaimant = (data) => {
  return (dispatch) => {
    dispatch(loadingClaimants());
    return axios
      .post(CLAIMANTS_API, data)
      .then(() => {
        dispatch(resetClaimants());
        dispatch(addSuccessNotification('Claimant created'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      });
  };
};

// action to update claimant by id
export const updateClaimant = (data) => {
  return (dispatch) => {
    dispatch(loadingClaimants());
    return axios
      .put(CLAIMANTS_API + '/' + data.id, data)
      .then((response) => {
        if (response.data.medium) dispatch(addMedia([response.data.medium]));
        response.data.description = {
          json: response.data.description,
          html: response.data.html_description,
        };
        dispatch(
          addClaimant(UPDATE_CLAIMANT, { ...response.data, medium: response.data.medium?.id }),
        );
        dispatch(addSuccessNotification('Claimant updated'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopClaimantsLoading()));
  };
};

// action to delete claimant by id
export const deleteClaimant = (id) => {
  return (dispatch) => {
    dispatch(loadingClaimants());
    return axios
      .delete(CLAIMANTS_API + '/' + id)
      .then(() => {
        dispatch(resetClaimants());
        dispatch(addSuccessNotification('Claimant deleted'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      });
  };
};

export const addClaimants = (claimants) => {
  return (dispatch) => {
    dispatch(
      addMedia(claimants.filter((claimant) => claimant.medium).map((claimant) => claimant.medium)),
    );
    dispatch(
      addClaimantsList(
        claimants.map((claimant) => {
          return { ...claimant, medium: claimant.medium?.id };
        }),
      ),
    );
  };
};

export const loadingClaimants = () => ({
  type: SET_CLAIMANTS_LOADING,
  payload: true,
});

export const stopClaimantsLoading = () => ({
  type: SET_CLAIMANTS_LOADING,
  payload: false,
});

export const addClaimant = (type, payload) => ({
  type,
  payload,
});

export const addClaimantsList = (data) => ({
  type: ADD_CLAIMANTS,
  payload: data,
});

export const addClaimantsRequest = (data) => ({
  type: ADD_CLAIMANTS_REQUEST,
  payload: data,
});

export const resetClaimants = () => ({
  type: RESET_CLAIMANTS,
});
