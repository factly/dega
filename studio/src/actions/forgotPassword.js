const userId = localStorage.getItem('userId');
export const requestPasswordReset = async (userId) => {
  const response = await fetch(
    `${window.REACT_APP_ZITADEL_AUTHORITY}/v2/users/${userId}/password_reset`,
    {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
        Authorization: `Bearer ${window.REACT_APP_ZITADEL_PAT}`,
      },
      body: JSON.stringify({
        sendLink: {
          notificationType: 'NOTIFICATION_TYPE_Email',
        },
      }),
    },
  );

  if (!response.ok) {
    throw new Error('Failed to send reset email');
  }

  return response.json();
};

export const resetPassword = async (userId, newPassword, verificationCode) => {
  const response = await fetch(
    `${window.REACT_APP_ZITADEL_AUTHORITY}/v2/users/${userId}/password`,
    {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
        Authorization: `Bearer ${window.REACT_APP_ZITADEL_PAT}`,
      },
      body: JSON.stringify({
        newPassword: {
          password: newPassword,
          changeRequired: false,
        },
        verificationCode: verificationCode,
      }),
    },
  );

  if (!response.ok) {
    throw new Error('Failed to reset password');
  }

  return response.json();
};
