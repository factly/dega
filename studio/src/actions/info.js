import axios from 'axios';
import { ADD_INFO, INFO_API, SET_INFO_LOADING } from '../constants/info';
import getError from '../utils/getError';
import { addErrorNotification } from './notifications';

export const getInfo = () => {
  return (dispatch) => {
    dispatch(loadingInfo(true));

    return axios
      .get(INFO_API)
      .then((response) => {
        dispatch(addInfo(response.data));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => {
        dispatch(loadingInfo(false));
      });
  };
};

export const addInfo = (data) => ({
  type: ADD_INFO,
  payload: data,
});

export const loadingInfo = (payload) => ({
  type: SET_INFO_LOADING,
  payload,
});
