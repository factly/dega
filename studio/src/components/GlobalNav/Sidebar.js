import React, { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Link, useLocation } from 'react-router-dom';
import { Layout, Menu, Popover, List, Avatar, Button } from 'antd';
import routes, { sidebarMenu } from '../../config/routesConfig';
import _ from 'lodash';
import { setCollapse } from './../../actions/sidebar';
import { setSpaceSelectorPage } from './../../actions/spaceSelectorPage';
import AccountMenu from './AccountMenu';
import {
  AppstoreOutlined,
  DownOutlined,
  MenuFoldOutlined,
  SettingOutlined,
  MenuUnfoldOutlined,
} from '@ant-design/icons';
import Search from '../Search';
import { maker } from '../../utils/sluger';
import degaImg from '../../assets/dega.png';

const { Sider } = Layout;
const { SubMenu } = Menu;

function Sidebar({ superOrg, permission, orgs, loading, applications, services, menuKey }) {
  const { collapsed } = useSelector((state) => state.sidebar);
  const dispatch = useDispatch();
  const { details, selected } = useSelector((state) => state.spaces);
  const { navTheme } = useSelector((state) => state.settings);
  const [showCoreMenu, setCoreMenu] = useState(false);
  const onCollapse = (collapsed) => {
    collapsed ? dispatch(setCollapse(true)) : dispatch(setCollapse(false));
  };

  if (loading) {
    return null;
  }
  let resource = [
    'home',
    'dashboard',
    'analytics',
    'google',
    'factly',
    'policies',
    'users',
    'spaces',
    'sach',
  ];
  let protectedResources = [
    'posts',
    'pages',
    'categories',
    'tags',
    'media',
    'formats',
    'claims',
    'claimants',
    'ratings',
    'organisations',
    'menus',
    'fact-checks',
    'episodes',
    'podcasts',
    'events',
    'webhooks',
    'website',
    'advanced',
    'members',
    'code-injection',
    'requests',
    'permissions',
  ];
  let buttonStyle = {
    width: '40px',
    height: '40px',
    background: '#DCE4E7',
    borderRadius: '21px',
    padding: '0.25rem 0.5rem',
  };
  permission.forEach((each) => {
    if (each.resource === 'admin') {
      resource = resource.concat(protectedResources);
      if (!showCoreMenu) {
        setCoreMenu(true);
      }
    } else {
      if (!showCoreMenu && protectedResources.includes(each.resource)) {
        setCoreMenu(true);
      }
      resource.push(each.resource);
    }
  });
  if (orgs[0]?.permission.role === 'owner') resource = resource.concat(protectedResources);
  const getMenuItems = (children, index, title) =>
    children.map((route, childIndex) => {
      return resource.includes(route.title.toLowerCase()) ? (
        ['Events', 'Permissions'].indexOf(route.title) !== -1 &&
        route.isAdmin !== superOrg.is_admin ? null : (
          <Menu.Item key={route.menuKey}>
            <Link to={route.path}>
              <span>{route.title}</span>
            </Link>
          </Menu.Item>
        )
      ) : null;
    });

  const getSubMenuItems = (menu, index, Icon) => (
    <SubMenu
      key={index}
      title={menu.title}
      icon={<Icon style={{ color: '#000', fontSize: '15px', fontWeight: '700' }} />}
    >
      {menu.submenu && menu.submenu.length > 0 ? (
        <>
          {menu.submenu.map((submenuItem, index) => {
            return orgs[0]?.permission.role === 'owner' ? (
              <SubMenu key={submenuItem.title + index} title={submenuItem.title}>
                {getMenuItems(submenuItem.children, index, submenuItem.title)}
              </SubMenu>
            ) : (
              <SubMenu key={submenuItem.title + index} title={submenuItem.title}>
                {getMenuItems(submenuItem.children, index, submenuItem.title)}
              </SubMenu>
            );
          })}
        </>
      ) : null}
      {getMenuItems(menu.children, index, menu.title)}
    </SubMenu>
  );

  return (
    <Sider
      breakpoint="xl"
      className="main-sidebar"
      width="264px"
      theme={navTheme}
      collapsible
      collapsed={collapsed}
      onCollapse={onCollapse}
      collapsedWidth={48}
      trigger={null}
      style={{
        position: 'sticky',
        background: '#f0f2f5',
        left: 0,
        top: 0,
        overflow: 'auto',
        height: '100vh',
      }}
    >
      <div
        style={{
          display: 'flex',
          flexDirection: 'column',
          justifyContent: 'space-between',
          alignItems: 'center',
          padding: collapsed ? '0 0.5rem' : '0 24px',
          position: 'sticky',
          top: 0,
          gap: '8px',
          height: collapsed && '100px',
          zIndex: 100,
          background: '#F1F1F1',
        }}
      >
        <div className="menu-header" style={{ width: '100%' }}>
          {collapsed ? (
            <img
              alt="logo"
              className="menu-logo"
              src={
                details[selected]?.fav_icon?.url?.[
                  window.REACT_APP_ENABLE_IMGPROXY ? 'proxy' : 'raw'
                ] || degaImg
              }
            />
          ) : (
            <Button
              style={{
                background: '#DCE4E7',
                width: '100%',
                color: '#1E1E1E',
                border: 'none',
                fontWeight: '600',
                fontSize: '1rem',
                display: 'flex',
                marginTop: '0.8rem',
                height: '43px',
                padding: '9px, 6px, 9px, 6px',
                justifyContent: 'space-between',
                alignItems: 'center',
              }}
              type="primary"
              onClick={() => dispatch(setSpaceSelectorPage(true))}
            >
              <div style={{ display: 'flex', alignItems: 'center', gap: '12px' }}>
                <Avatar
                  src={
                    details[selected]?.fav_icon?.url?.[
                      window.REACT_APP_ENABLE_IMGPROXY ? 'proxy' : 'raw'
                    ] || degaImg
                  }
                />
                {details[selected]?.name}
              </div>
              <DownOutlined />
            </Button>
          )}
        </div>
        <Search collapsed={collapsed} />
      </div>
      <Menu
        theme={navTheme}
        mode="inline"
        className="slider-menu"
        defaultOpenKeys={['0', '1']}
        style={{ background: '#F1F1F1', padding: '8px' }}
        selectedKeys={menuKey}
      >
        {sidebarMenu.map((menu, index) => {
          const { Icon } = menu;
          return menu.title === 'CORE' && !showCoreMenu
            ? null
            : !menu.isService
            ? !menu.isAdmin
              ? getSubMenuItems(menu, index, Icon)
              : permission.filter((each) => each.resource === 'admin').length > 0
              ? getSubMenuItems(menu, index, Icon)
              : null
            : services?.includes(maker(menu.title))
            ? getSubMenuItems(menu, index, Icon)
            : null;
        })}
      </Menu>
      {!collapsed ? (
        <div
          style={{
            display: 'flex',
            justifyContent: 'space-between',
            padding: '12px 24px',
            lineHeight: '40px',
            alignItems: 'center',
            width: '100%',
            position: 'absolute',
            bottom: '0',
            background: '#f0f2f5',
          }}
        >
          <AccountMenu />
          <div>
            <Link to="/settings">
              <Button style={{ ...buttonStyle }}>
                <SettingOutlined />
              </Button>
            </Link>
            {applications.length > 0 ? (
              <>
                <Popover
                  placement="top"
                  overlayInnerStyle={{ paddingBottom: 0 }}
                  content={
                    <List
                      grid={{
                        gutter: 16,
                        column: 3,
                      }}
                      dataSource={applications}
                      renderItem={(item) => (
                        <List.Item>
                          <a
                            href={item.url}
                            style={{
                              margin: '8px',
                              textDecoration: 'none',
                              color: 'inherit',
                              display: 'flex',
                              flexDirection: 'column',
                              justifyContent: 'center',
                              alignItems: 'center',
                            }}
                          >
                            {item.medium && item.medium.url ? (
                              <img alt="logo" className="menu-logo" src={item.medium.url.raw} />
                            ) : (
                              <Avatar shape="square" size={35}>
                                {item.name.charAt(0)}
                              </Avatar>
                            )}
                            <span>{item.name}</span>
                          </a>
                        </List.Item>
                      )}
                    />
                  }
                  trigger="click"
                >
                  <Button style={{ ...buttonStyle, marginLeft: '0.25rem' }}>
                    <AppstoreOutlined />
                  </Button>
                </Popover>
              </>
            ) : null}
            <Button
              style={{ ...buttonStyle, marginLeft: '0.25rem' }}
              onClick={() => onCollapse(true)}
            >
              <MenuFoldOutlined />
            </Button>
          </div>
        </div>
      ) : (
        <div
          style={{
            display: 'flex',
            justifyContent: 'space-between',
            padding: '12px 8px',
            lineHeight: '40px',
            alignItems: 'center',
            width: '100%',
            position: 'absolute',
            bottom: '0',
            background: '#f0f2f5',
          }}
        >
          <Button style={buttonStyle} onClick={() => onCollapse(false)}>
            <MenuUnfoldOutlined />
          </Button>
        </div>
      )}
    </Sider>
  );
}
export default Sidebar;
