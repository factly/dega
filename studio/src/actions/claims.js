import axios from 'axios';
import {
  ADD_CLAIMS,
  ADD_CLAIMS_REQUEST,
  SET_CLAIMS_LOADING,
  RESET_CLAIMS,
  CLAIMS_API,
  GET_CLAIM,
  UPDATE_CLAIM,
} from '../constants/claims';
import { addErrorNotification, addSuccessNotification } from './notifications';
import { addRatings } from './ratings';
import { addClaimants } from './claimants';
import getError from '../utils/getError';

// action to fetch all claims
export const getClaims = (query) => {
  const params = new URLSearchParams();
  if (query.claimant && query.claimant.length > 0) {
    query.claimant.map((each) => params.append('claimant', each));
  }
  if (query.rating && query.rating.length > 0) {
    query.rating.map((each) => params.append('rating', each));
  }
  if (query.page) {
    params.append('page', query.page);
  }
  if (query.limit) {
    params.append('limit', query.limit);
  }
  if (query.sort) {
    params.append('sort', query.sort);
  }
  if (query.q) {
    params.append('q', query.q);
  }
  return (dispatch) => {
    dispatch(loadingClaims());
    return axios
      .get(CLAIMS_API, {
        params: params,
      })
      .then((response) => {
        dispatch(
          addClaimants(
            response.data.nodes.filter((claim) => claim.claimant).map((claim) => claim.claimant),
          ),
        );
        dispatch(
          addRatings(
            response.data.nodes.filter((claim) => claim.rating).map((claim) => claim.rating),
          ),
        );
        dispatch(
          addClaimsList(
            response.data.nodes.map((claim) => {
              claim.description = { json: claim.description, html: claim.description_html };
              return { ...claim, claimant: claim.claimant.id, rating: claim.rating.id };
            }),
          ),
        );
        dispatch(
          addClaimsRequest({
            data: response.data.nodes.map((item) => item.id),
            query: query,
            total: response.data.total,
          }),
        );
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopClaimsLoading()));
  };
};

// action to fetch claim by id
export const getClaim = (id) => {
  return (dispatch) => {
    dispatch(loadingClaims());
    return axios
      .get(CLAIMS_API + '/' + id)
      .then((response) => {
        let claim = response.data;
        claim.description = { json: claim.description, html: claim.description_html };
        dispatch(addClaimants([claim.claimant]));
        dispatch(addRatings([claim.rating]));

        dispatch(
          addClaim(GET_CLAIM, { ...claim, claimant: claim.claimant.id, rating: claim.rating.id }),
        );
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopClaimsLoading()));
  };
};

// action to create claim
export const createClaim = (data) => {
  return (dispatch) => {
    dispatch(loadingClaims());
    return axios
      .post(CLAIMS_API, data)
      .then((response) => {
        let claim = response.data;
        claim.description = { json: claim.description, html: claim.description_html };
        dispatch(addClaimants([claim.claimant]));
        dispatch(addRatings([claim.rating]));

        dispatch(resetClaims());
        dispatch(addSuccessNotification('Claim created'));
        return claim;
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      });
  };
};

// action to update claim by id
export const updateClaim = (data) => {
  return (dispatch) => {
    dispatch(loadingClaims());
    return axios
      .put(CLAIMS_API + '/' + data.id, data)
      .then((response) => {
        let claim = response.data;
        claim.description = { json: claim.description, html: claim.description_html };
        dispatch(addClaimants([claim.claimant]));
        dispatch(addRatings([claim.rating]));

        dispatch(
          addClaim(UPDATE_CLAIM, {
            ...claim,
            claimant: claim.claimant.id,
            rating: claim.rating.id,
          }),
        );
        dispatch(addSuccessNotification('Claim updated'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      })
      .finally(() => dispatch(stopClaimsLoading()));
  };
};

// action to delete claim by id
export const deleteClaim = (id) => {
  return (dispatch) => {
    dispatch(loadingClaims());
    return axios
      .delete(CLAIMS_API + '/' + id)
      .then(() => {
        dispatch(resetClaims());
        dispatch(addSuccessNotification('Claim deleted'));
      })
      .catch((error) => {
        dispatch(addErrorNotification(getError(error)));
      });
  };
};

export const addClaims = (claims) => {
  return (dispatch) => {
    dispatch(addClaimants(claims.filter((claim) => claim.claimant).map((claim) => claim.claimant)));
    dispatch(addRatings(claims.filter((claim) => claim.rating).map((claim) => claim.rating)));
    dispatch(
      addClaimsList(
        claims.map((claim) => {
          return { ...claim, claimant: claim.claimant.id, rating: claim.rating.id };
        }),
      ),
    );
  };
};

export const loadingClaims = () => ({
  type: SET_CLAIMS_LOADING,
  payload: true,
});

export const stopClaimsLoading = () => ({
  type: SET_CLAIMS_LOADING,
  payload: false,
});

export const addClaim = (type, payload) => ({
  type,
  payload,
});

export const addClaimsList = (payload) => ({
  type: ADD_CLAIMS,
  payload,
});

export const addClaimsRequest = (payload) => ({
  type: ADD_CLAIMS_REQUEST,
  payload,
});

export const resetClaims = () => ({
  type: RESET_CLAIMS,
});
