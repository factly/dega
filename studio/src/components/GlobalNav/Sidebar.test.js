import React from 'react';
import { useHistory, useLocation } from 'react-router-dom';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { useDispatch, Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount, shallow } from 'enzyme';

import '../../matchMedia.mock';
import Sidebar from './Sidebar';
import * as actions from '../../actions/sidebar';
import { Layout } from 'antd';

const { Sider } = Layout;

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
}));

jest.mock('react-monaco-editor', () => {
  const MonacoEditor = () => <div />;
  return MonacoEditor;
});

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: jest.fn(),
  useLocation: () => ({
    pathname: '/posts',
  }),
}));

jest.mock('../../actions/sidebar', () => ({
  setCollapse: jest.fn(),
}));

describe('Sidebar component', () => {
  let store;
  let mockedDispatch;

  let state = {
    settings: {
      navTheme: 'dark',
      title: 'Dega',
      sider: {
        collapsed: true,
      },
    },
    sidebar: {
      collapsed: true,
    },
    spaces: {
      orgs: [{ id: 1, spaces: [1] }],
      details: {
        1: {
          id: 1,
          created_at: '2020-09-23T06:11:32.986694Z',
          updated_at: '2020-09-23T06:11:32.986694Z',
          name: 'English',
          slug: 'english',
          organisation_id: 1,
          permissions: [
            {
              resource: 'posts',
              actions: ['create'],
            },
          ],
        },
      },
      loading: false,
      selected: 1,
    },
    search: {
      req: [],
      details: {
        articles: [],
        'fact-checks': [],
        pages: [],
        claims: [],
        tags: [],
        categories: [],
        media: [],
        ratings: [],
        total: 0,
      },
      loading: true,
    },
    admin: {
      organisation: {
        id: 1,
        is_admin: true,
      },
    },
  };
  store = mockStore(() => state);
  store.dispatch = jest.fn(() => ({}));
  mockedDispatch = jest.fn(() => Promise.resolve({}));
  useDispatch.mockReturnValue(mockedDispatch);
  describe('snapshot testing', () => {
    it('should render the component', () => {
      const tree = shallow(
        <Provider store={store}>
          <Router>
            <Sidebar
              services={['core', 'fact-checking', 'podcast']}
              permission={[{ resource: 'posts', actions: ['create'] }]}
              orgs={[{ id: 1, spaces: [1], permission: { role: 'owner' } }]}
            />
          </Router>
        </Provider>,
      );
      expect(tree).toMatchSnapshot();
    });
    it('should not render the component when no space', () => {
      store = mockStore({
        settings: state.settings,
        sidebar: { collapsed: true },
        spaces: { selected: 0 },
      });
      const tree = shallow(
        <Provider store={store}>
          <Router>
            <Sidebar
              services={['core', 'fact-checking', 'podcast']}
              permission={[{ resource: 'posts', actions: ['create'] }]}
              orgs={[{ id: 1, spaces: [], permission: { role: 'owner' } }]}
            />
          </Router>
        </Provider>,
      );
      expect(tree).toMatchSnapshot();
    });
  });
  describe('component testing', () => {
    let wrapper;
    it('should call toggleSider', (done) => {
      actions.setCollapse.mockReset();
      const push = jest.fn();
      useHistory.mockReturnValueOnce({ push });
      store = mockStore(state);

      wrapper = mount(
        <Provider store={store}>
          <Router>
            <Sidebar
              services={['core', 'fact-checking', 'podcast']}
              permission={[{ resource: 'posts', actions: ['create'] }]}
              orgs={[{ id: 1, spaces: [1], permission: { role: 'owner' } }]}
            />
          </Router>
        </Provider>,
      );

      wrapper.find(Sider).props().onCollapse();
      wrapper.update();

      expect(actions.setCollapse).toHaveBeenCalled();
      done();
    });
    it('should work alos when resource is admin', (done) => {
      actions.setCollapse.mockReset();
      const push = jest.fn();
      useHistory.mockReturnValueOnce({ push });
      store = mockStore(state);

      wrapper = mount(
        <Provider store={store}>
          <Router>
            <Sidebar
              services={['core', 'fact-checking', 'podcast']}
              permission={[{ resource: 'admin' }]}
              orgs={[{ id: 1, spaces: [1], permission: { role: 'owner' } }]}
            />
          </Router>
        </Provider>,
      );

      wrapper.find(Sider).props().onCollapse();
      wrapper.update();

      expect(actions.setCollapse).toHaveBeenCalled();
      done();
    });
    it('should work alos when resource is not admin', (done) => {
      actions.setCollapse.mockReset();
      const push = jest.fn();
      useHistory.mockReturnValueOnce({ push });
      store = mockStore(state);

      wrapper = mount(
        <Provider store={store}>
          <Router>
            <Sidebar
              services={['core', 'fact-checking', 'podcast']}
              permission={[{ resource: 'posts', actions: ['update'] }]}
              orgs={[{ id: 1, spaces: [1], permission: { role: 'member' } }]}
            />
          </Router>
        </Provider>,
      );

      wrapper.find(Sider).props().onCollapse();
      wrapper.update();

      expect(actions.setCollapse).toHaveBeenCalled();
      done();
    });
    it('should not display when loading', (done) => {
      actions.setCollapse.mockReset();
      const push = jest.fn();
      useHistory.mockReturnValueOnce({ push });
      store = mockStore(state);

      wrapper = mount(
        <Provider store={store}>
          <Router>
            <Sidebar
              services={['core', 'fact-checking', 'podcast']}
              loading={true}
              permission={[{ resource: 'admin' }]}
              orgs={[{ id: 1, spaces: [1], permission: { role: 'owner' } }]}
            />
          </Router>
        </Provider>,
      );
      expect(wrapper.find(Sider).length).toBe(0);
      expect(actions.setCollapse).not.toHaveBeenCalled();
      done();
    });
  });
});
