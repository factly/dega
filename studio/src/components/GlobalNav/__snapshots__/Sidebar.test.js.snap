// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar component snapshot testing should not render the component when no space 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <BrowserRouter>
    <Sidebar
      orgs={
        Array [
          Object {
            "id": 1,
            "permission": Object {
              "role": "owner",
            },
            "spaces": Array [],
          },
        ]
      }
      permission={
        Array [
          Object {
            "actions": Array [
              "create",
            ],
            "resource": "posts",
          },
        ]
      }
      services={
        Array [
          "core",
          "fact-checking",
          "podcast",
        ]
      }
      superOrg={
        Object {
          "id": 1,
          "is_admin": true,
        }
      }
    />
  </BrowserRouter>
</ContextProvider>
`;

exports[`Sidebar component snapshot testing should render the component 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [MockFunction],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [MockFunction],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <BrowserRouter>
    <Sidebar
      orgs={
        Array [
          Object {
            "id": 1,
            "permission": Object {
              "role": "owner",
            },
            "spaces": Array [
              1,
            ],
          },
        ]
      }
      permission={
        Array [
          Object {
            "actions": Array [
              "create",
            ],
            "resource": "posts",
          },
        ]
      }
      services={
        Array [
          "core",
          "fact-checking",
          "podcast",
        ]
      }
      superOrg={
        Object {
          "id": 1,
          "is_admin": true,
        }
      }
    />
  </BrowserRouter>
</ContextProvider>
`;
